grudge/dt_utils.py

```diff
@@ -271,31 +271,52 @@ def dt_geometric_factors(
     dd_face = DOFDesc("all_faces", dd.discretization_tag)
     face_discr = dcoll.discr_from_dd(dd_face)
 
-    # To get a single value for the total surface area of a cell, we
-    # take the sum over the averaged face areas on each face.
-    # NOTE: The face areas are the *same* at each face nodal location.
-    # This assumes there are the *same* number of face nodes on each face.
     surface_areas = abs(
         op.elementwise_integral(
             dcoll, dd_face, face_discr.zeros(actx) + 1.0
         )
     )
-    surface_areas = DOFArray(
-        actx,
-        data=tuple(
-            actx.einsum("fej->e",
-                        face_ae_i.reshape(
-                            vgrp.mesh_el_group.nfaces,
-                            vgrp.nelements,
-                            afgrp.nunit_dofs
-                        ),
-                        tagged=(FirstAxisIsElementsTag(),)) / afgrp.nunit_dofs
-
-            for vgrp, afgrp, face_ae_i in zip(volm_discr.groups,
-                                              face_discr.groups,
-                                              surface_areas)
+
+    if actx.supports_nonscalar_broadcasting:
+        # Sum the areas of each face to get the total surface area
+        surface_areas = DOFArray(
+            actx,
+            data=tuple(
+                actx.einsum(
+                    "fej->e",
+                    face_ae_i.reshape(
+                        vgrp.mesh_el_group.nfaces,
+                        vgrp.nelements,
+                        1
+                    ),
+                    tagged=(FirstAxisIsElementsTag(),))
+
+                for vgrp, face_ae_i in zip(volm_discr.groups, surface_areas)
+            )
+        )
+    else:
+        # Whenever the array context can't perform nonscalar broadcasting,
+        # elementwise reductions (like `elementwise_integral`) repeat the
+        # scalar value of the reduction at each degree of freedom.
+        # To get a single value for the total surface area of a cell,
+        # we take the sum over the averaged face areas on each face.
+        surface_areas = DOFArray(
+            actx,
+            data=tuple(
+                actx.einsum(
+                    "fej->e",
+                    face_ae_i.reshape(
+                        vgrp.mesh_el_group.nfaces,
+                        vgrp.nelements,
+                        afgrp.nunit_dofs
+                    ),
+                    tagged=(FirstAxisIsElementsTag(),)) / afgrp.nunit_dofs
+
+                for vgrp, afgrp, face_ae_i in zip(volm_discr.groups,
+                                                  face_discr.groups,
+                                                  surface_areas)
+            )
```

Comment:
These both branches could be integrated if we called
```python
face_ae_i.reshape(
    vgrp.mesh_el_group.nfaces,
    vgrp.nelements,
    -1
)
```

Reply:
Good call --- fixed: https://github.com/inducer/grudge/pull/195/commits/db98f0dff37cccef7f8d3216057487a4456171fc